Fixed : Comment notification settings are visible even if comments ar…

tusharnain opened the pull request:
…e disabled

Added a UX condition to display comment notification settings, only if the user has enabled the comment notifications.

tusharnain committed:
Fixed : Comment notification settings are visible even if comments are disabled

Added a UX condition to display comment notification settings, only if the user has enabled the comment notifications.

ssddanbrown commented:
Thanks for offering this @tusharnain4578,

Please could you remove the changes from 718adf38a0c14e037e9e90bf328875fd5a53110d from this PR?
They appear unrelated and cause many changes to show, I'd want this PR to be focused on the single issue (Plus I would not accept a PR at this time to change default credentials).

Also, would you be willing to add a test case to cover this functionality? Just something to add to our existing [UserPreference tests](https://github.com/BookStackApp/BookStack/blob/dfaf6f7c13e5cee904a43f9e8ad3777ca512fac0/tests/User/UserPreferencesTest.php#L108).

tusharnain commented:
@ssddanbrown Sorry for your troubles, i have removed the unnecessary commit from the PR, but i couldn't get what you mean by 'add a testcase', could you please elaborate more, sir?
And also i am sorry for any inconvinience and stupidity from my side, i am very new to open source, git and all.

ssddanbrown commented:
@tusharnain4578 No problem, no need to apologise, everyone is new to things at one point, and a lot of things kind of thing can be project specific.

> but i couldn't get what you mean by 'add a testcase', could you please elaborate more, sir?

We have a big set of automated tests that specific things, which is ran via phpunit. 
For each bit of functionality we fix/change (at least for back-end code), I like to try to cover it with a test to prevent the issue coming back in the future.
Here's an example of a single test:

https://github.com/BookStackApp/BookStack/blob/dfaf6f7c13e5cee904a43f9e8ad3777ca512fac0/tests/User/UserPreferencesTest.php#L108-L125

This loads the `/preferences/notifications' view, then checks the response to ensure it has the content we expect (In that case, a book we're watching). 
For this, I'd like something similar which checks that the options are not shown if comments are disabled.

There's a bit of guidance on this [in our devdocs here](https://github.com/BookStackApp/BookStack/blob/9ac932fc28ac7085eb956c2c217e9b8612462d4a/dev/docs/development.md?plain=1#L26-L30) although I've just noticed it's not very well documented and under the wrong heading, I'll look to address that at some point.

If it's too much to learn/understand/get-involved-with just let me know and I can instead add the testing myself during review and merge.

tusharnain committed:
Update UserPreferencesTest.php

Added Testcases for preferences menu of Comment Notifications visibility when comments are enabled/disabled.

tusharnain4578 commented:
I have added 2 test cases for this functionality, please review, sir.
Here #4552.

tusharnain commented:
I have added 2 test cases in UserPreferencesTest.php for this functionality, please review, sir.

ssddanbrown commented:
Thanks @tusharnain4578, now merged into development to be part of the next feature release.
I followed this up with c42cd29ed3e2fe1eca4c89e679197b76db699e03 during review, just to make some tweaks to the test.

ssddanbrown commented:
Following on from my comment yesterday, regarding our bad documentation for PHP tests, I've now updated and expanded the document for this into it's own document:

https://github.com/BookStackApp/BookStack/blob/development/dev/docs/php-testing.md

igoresz commented:
Sorry guys for maybe abuse asking... Notifications in emails by new comments on WATCHed pages - is it in perspective roadmap?
Many thanks in advance. Good project, by the way! We use it, it helps us much.

Man-in-Black commented:
when you click on the watching setting, is that what you're searching for?
![image](https://github.com/BookStackApp/BookStack/assets/9048534/c133f165-a2a2-4ab6-90b8-72b6a449a1ff)
here you can simply change the level of notification to also include mails on comments.
